File: agentloop/truncation.go
package agentloop

import (
	"fmt"
	"strings"
)

// TruncationMode specifies how output is truncated.
type TruncationMode string

const (
	TruncateHeadTail TruncationMode = "head_tail"
	TruncateTail     TruncationMode = "tail"
)

// Default character limits per tool (Section 5.2).
var DefaultToolCharLimits = map[string]int{
	"read_file":   50000,
	"shell":       30000,
	"grep":        20000,
	"glob":        20000,
	"edit_file":   10000,
	"apply_patch": 10000,
	"write_file":  1000,
	"spawn_agent": 20000,
}

// Default truncation modes per tool.
var DefaultTruncationModes = map[string]TruncationMode{
	"read_file":   TruncateHeadTail,
	"shell":       TruncateHeadTail,
	"grep":        TruncateTail,
	"glob":        TruncateTail,
	"edit_file":   TruncateTail,
	"apply_patch": TruncateTail,
	"write_file":  TruncateTail,
	"spawn_agent": TruncateHeadTail,
}

// Default line limits per tool (applied after character truncation).
var DefaultToolLineLimits = map[string]int{
	"shell": 256,
	"grep":  200,
	"glob":  500,
}

// TruncateOutput applies character-based truncation to output.
func TruncateOutput(output string, maxChars int, mode TruncationMode) string {
	if len(output) <= maxChars {
		return output
	}

	switch mode {
	case TruncateHeadTail:
		half := maxChars / 2
		removed := len(output) - maxChars
		return output[:half] +
			fmt.Sprintf("\n\n[WARNING: Tool output was truncated. %d characters were removed from the middle. "+
				"The full output is available in the event stream. "+
				"If you need to see specific parts, re-run the tool with more targeted parameters.]\n\n",
				removed) +
			output[len(output)-half:]

	case TruncateTail:
		removed := len(output) - maxChars
		return fmt.Sprintf("[WARNING: Tool output was truncated. First %d characters were removed. "+
			"The full output is available in the event stream.]\n\n",
			removed) +
			output[len(output)-maxChars:]

	default:
		// Default to head_tail.
		half := maxChars / 2
		removed := len(output) - maxChars
		return output[:half] +
			fmt.Sprintf("\n\n[WARNING: Tool output was truncated. %d characters were removed from the middle.]\n\n",
				removed) +
			output[len(output)-half:]
	}
}

// TruncateLines applies line-based truncation using head/tail split.
func TruncateLines(output string, maxLines int) string {
	lines := strings.Split(output, "\n")
	if len(lines) <= maxLines {
		return output
	}

	headCount := maxLines / 2
	tailCount := maxLines - headCount
	omitted := len(lines) - headCount - tailCount

	return strings.Join(lines[:headCount], "\n") +
		fmt.Sprintf("\n[... %d lines omitted ...]\n", omitted) +
		strings.Join(lines[len(lines)-tailCount:], "\n")
}

// TruncateToolOutput applies the full truncation pipeline for a tool:
// 1. Character-based truncation (primary, handles pathological cases)
// 2. Line-based truncation (secondary, for readability)
func TruncateToolOutput(output string, toolName string, charLimits map[string]int, lineLimits map[string]int) string {
	// Step 1: Character-based truncation.
	maxChars, ok := charLimits[toolName]
	if !ok {
		maxChars, ok = DefaultToolCharLimits[toolName]
		if !ok {
			maxChars = 30000 // fallback default
		}
	}

	mode, ok := DefaultTruncationModes[toolName]
	if !ok {
		mode = TruncateHeadTail
	}

	result := TruncateOutput(output, maxChars, mode)

	// Step 2: Line-based truncation.
	maxLines := 0
	if lineLimits != nil {
		if ml, ok := lineLimits[toolName]; ok {
			maxLines = ml
		}
	}
	if maxLines == 0 {
		if ml, ok := DefaultToolLineLimits[toolName]; ok {
			maxLines = ml
		}
	}
	if maxLines > 0 {
		result = TruncateLines(result, maxLines)
	}

	return result
}

File: packages/attractor-agent/src/attractor_agent/truncation.py
"""Tool output truncation helpers."""

from __future__ import annotations

DEFAULT_TOOL_CHAR_LIMITS = {
    "read_file": 50_000,
    "shell": 30_000,
    "grep": 20_000,
    "glob": 20_000,
    "edit_file": 10_000,
    "apply_patch": 10_000,
    "write_file": 1_000,
}

DEFAULT_TOOL_LINE_LIMITS = {
    "shell": 256,
    "grep": 200,
    "glob": 500,
}

DEFAULT_TOOL_MODES = {
    "read_file": "head_tail",
    "shell": "head_tail",
    "grep": "tail",
    "glob": "tail",
    "edit_file": "tail",
    "apply_patch": "tail",
    "write_file": "tail",
}


def truncate_output(output: str, max_chars: int, mode: str = "head_tail") -> str:
    if max_chars <= 0 or len(output) <= max_chars:
        return output

    if mode == "tail":
        removed = len(output) - max_chars
        return (
            "[WARNING: Tool output was truncated. First "
            f"{removed} characters were removed. The full output is available in the event stream.]\n\n"
            f"{output[-max_chars:]}"
        )

    head_chars = max_chars // 2
    tail_chars = max_chars - head_chars
    removed = len(output) - max_chars
    return (
        f"{output[:head_chars]}\n\n"
        "[WARNING: Tool output was truncated. "
        f"{removed} characters were removed from the middle. "
        "The full output is available in the event stream. "
        "If you need to see specific parts, re-run the tool with more targeted parameters.]\n\n"
        f"{output[-tail_chars:]}"
    )


def truncate_lines(output: str, max_lines: int) -> str:
    if max_lines <= 0:
        return output

    lines = output.splitlines()
    if len(lines) <= max_lines:
        return output

    head_count = max_lines // 2
    tail_count = max_lines - head_count
    omitted = len(lines) - head_count - tail_count
    kept = lines[:head_count] + [f"[... {omitted} lines omitted ...]"] + lines[-tail_count:]
    if "[WARNING:" in output and not any(line.startswith("[WARNING:") for line in kept):
        warning_line = next((line for line in lines if line.startswith("[WARNING:")), None)
        if warning_line is not None:
            kept.insert(0, warning_line)
    return "\n".join(kept)


def truncate_tool_output(
    output: str,
    tool_name: str,
    tool_char_limits: dict[str, int] | None = None,
    tool_line_limits: dict[str, int] | None = None,
) -> str:
    char_limits = DEFAULT_TOOL_CHAR_LIMITS | (tool_char_limits or {})
    line_limits = DEFAULT_TOOL_LINE_LIMITS | (tool_line_limits or {})
    mode = DEFAULT_TOOL_MODES.get(tool_name, "head_tail")
    max_chars = char_limits.get(tool_name, 10_000)

    result = truncate_output(output, max_chars=max_chars, mode=mode)
    max_lines = line_limits.get(tool_name)
    if max_lines is not None:
        result = truncate_lines(result, max_lines=max_lines)
    return result
